### DISCUSSION TITLE
Extending an existing endpoint

### DISCUSSION THREAD

**Discussion by WildsG** [medusajs/medusa#8473] [Q&A]
Hi. 

Some background what I need to achieve.
I want to let user to upload an image besides a product when user is adding it to cart. 
So the problem I need help with, is how to properly customize medusa backend and storefront to be able to achieve this. 

Main issues are:
1. In medusa admin, I can't really extend an endpoint. However, I can create a new one, but I need all the same functionality as an old endpoint, just some tweaks while persiting the data. There's no easy way to do this. What I did - I went to github, found the endpoints source code, copied it and tried to add some of my code (this kinda works, but the process to do this seems way to tedious).
2. In Storefront (frontend starter application provided by medusajs), now I need to call this newly implemented endpoint with all the same parameters for the request, just need some new properties added to the request object. Since storefront uses medusaClient everywhere, I can't really take request code and tweak it for my needs. Furthermore, I can't find the implementation of medusa-js library anywhere to be able to do what I did with backend - just copy existing implementation and change it or add my custom code. So how would I do that? 

For example, how can I change the following requests path - I need it to call my new endpoint in a new route. Also, how could I add some new properties to request object? Since all of this functionality is embedded into library, I can't get any good reference point to replicate existing functionality. 

![image](https://github.com/user-attachments/assets/a7987927-158f-4359-a463-d8fc99bd8c15)

**Comment by WildsG** [comment]
To follow up. 

To achieve what I'm trying to achieve, having new endpoint is not neccessary. However, then I need a way to persist data into additional fields of an entity.  How would I do that from start to finish - from sending a request to the server and actually storing that data after some transformations into the database? Should I use medusa widgets for that?

P.S. 
After reading some more into the other discussions, it really seems that a lot of these basic processes are not documented well, or not documented at all and users, such as myself are really strugling :/

**Comment by loki-dog** [comment]
I can't say on the full process, but if you want to extend the endpoint you can extend whichever entity you want: https://docs.medusajs.com/development/entities/extend-entity and same for the service. 

But if you want to have less exposure to doing everything yourself, could also consider the medusa plugin medusa-extender https://github.com/adrien2p/medusa-extender. I've not used it but it claims to make the process of extending existing endpoints and creating new ones easy.

edit to say that you didn't mention storage. But you will need to consider what to do with the image, you can store raw image data in the database but that can get bloat your DB quickly and it can be messy unless you have a well-designed system.  The most likely scenario is upload the user's file to S3 or other storage, and just persist the link in medusa by extending the revelent entity and service. Frontend can fetch from the storage which reduces your server load. There are plugins for storage if you don't want to write your own :)

**Reply by WildsG** [reply]
Hi @loki-dog . Thanks for the response. 
Thanks for mentioning image storage, but I've already got all of that covered. I will be indeed using s3 :)

So to extend an entity is not a problem. 
However, the issue is extending an endpoint. I want to add additional processing (in this case, as you said, uploading my image to S3). 
So what I figured I would do, was to add a new endpoint, having old endpoint functionality and before also mine (image uploading to S3). 
So as I mentioned, I managed to find a way to actually extend an endpoint (how I described above). However, I now need to call this new route in my frontend.
Frontend application uses medusa-js, and it appears that there is no public source code for it, and I can't mimick the request it was doing to the original endpoint (I could by reverse engineering what I see in developer console, but this is just insane). But other than that there is no way to do something like that. 

MedusaExtender by the way is deprecated :/ 

[In the extend-entity](https://docs.medusajs.com/development/entities/extend-entity) 
This link shows how to add additional field, but it does not explain how to add some processing in the backend to fill that field. Only how to send its contents from storefront. So this does not work for me :(

Any other ideas?
